File: Courtney/CCSubsystemCodeforVehicle/DHKE/DHKE.cpp
#include <iostream>
#include <cmath>

#define MAX 100000
typedef long long ll;
using namespace std;

int main(){
    //##### Diffie-Hellman Key Exchange Before Going to Device #####
    //Initializing P and G
    int P;
    P = 13;
    int G;
    G = 7;
    long long randInt;

    //Generating a random number B
    srand((unsigned) time(NULL));
    long long b = 1 + (rand() % 15);

    //Doing the Diffie-Hellman Mathematics - G^b mod P
    long long Bprime;
    Bprime = static_cast<long long>(pow(G, b)) % P;

    //Sending the value to the device - outputting for demonstration purposes
    std::cout << "B Value to Device: " << Bprime << "\n";
    
    //##### Diffie-Hellman Key Exchange After Going to Vehicle #####
    //Getting the value from the Device
    long long A;
    std::cout << "Type in A from Device: ";
    std::cin >> A;
    
    //Doing the Diffie-Hellman Mathematics - A^b mod P
    long long sk;
    sk = static_cast<long long>(pow(A, b)) % P;
    
    //Outputting the final number for demonstation purposes
    std::cout << "Secret Key: " << sk << "\n";

    return 0;
}

File: DigitalKeyFOBcomp/app/src/main/java/com/example/digitalkeyfobcomp/Encryption/DHKE.kt
package com.example.digitalkeyfobcomp.Encryption

import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import java.math.BigInteger
import java.security.SecureRandom

fun twoPair(B: BigInteger, a: Int): Pair<BigInteger, Int> = B to a // equivalent to Pair("Ali", 33)

fun DHKEBefore(): Pair<BigInteger, Int> {
//    Surface(color = Color.LightGray) {
        //Setting P and G - P is the prime number and G is the generator (primative root of P)
        val P = BigInteger("13")
        val G = BigInteger("7")

        //Creating the random A for the Vehicle
        val randInt = SecureRandom() //special randomization that follow standard cryptography protocols
        val a = randInt.nextInt(15) // pass to DHKE after

        //Doing the Diffie-Hellman Process (G^a mod P)
        var A = BigInteger("2")
        A = (G.pow(a)) % P //mathematics

        val retunval = twoPair(A,a)
        return retunval
        //Outputs the text to use for demonstration and debugging purposes
//        Text(text = "Before Vehicle: P = $P \n\nG = $G" +
//                "\n\nA Val = $a" +
//                "\n\nVehicle DHKE Portion = $A",
//            modifier = modifier.padding(12.dp))
//    }
}

fun DHKEAfter(B: BigInteger, a: Int): BigInteger {
//    Surface(color = Color.LightGray) {
        //Setting the Values of P, A, and B
        //P is predetermined, A is calculated before, and B is taken from the vehicle.
        val P = BigInteger("13")
//        val a: Int = 1 //This has to be changed with the unique A value from the first part.

//        val B = BigInteger("7") //Add in the value of B from the vehicle here.

        //Calculating the final secret key - B^a mod P
        val sk = (B.pow(a) % P)
        //Text Output for demonstration and debugging.
//        Text(text = "Testing: \nSk = $sk", modifier = modifier.padding(12.dp))
        return sk
}